src/Agent.Worker/Build/SourceProvider.cs

```diff
@@ -15,6 +16,10 @@ public interface ISourceProvider : IExtension, IAgentService
         string RepositoryType { get; }
 
         string GetBuildDirectoryHashKey(IExecutionContext executionContext, ServiceEndpoint endpoint);
+
+        Task GetSourceAsync(IExecutionContext executionContext, ServiceEndpoint endpoint, CancellationToken cancellationToken);
+
+        Task ResetRepositoryAsync(IExecutionContext executionContext, ServiceEndpoint endpoint);
```

Comment:
I am not able to come up with a good name. the problem is this method will be used across all repository type, git/tfvc/svn.